### PR TITLE
Reload reclaimed bloom filters when memory is available

### PR DESCRIPTION
**Pull request by lkshminarayanan**
PR #17771 introduced a threshold for the total memory used by all bloom filters across SSTables. When the total usage surpasses the threshold, the largest bloom filter will be removed from memory, bringing the total usage back under the threshold. This PR adds support for reloading such reclaimed bloom filters back into memory when memory becomes available (i.e., within the 10% of available memory earmarked for the reclaimable components). 

The SSTables manager now maintains a list of all SSTables whose bloom filter was removed from memory and attempts to reload them when an SSTable, whose bloom filter is still in memory, gets deleted. The manager reloads from the smallest to the largest bloom filter to maximize the number of filters being reloaded into memory.

### CODE REVIEW

File: sstables/sstables.cc [hunk 1]
```diff
@@ -1408,20 +1408,25 @@ size_t sstable::total_reclaimable_memory_size() const {
 }
 
 size_t sstable::reclaim_memory_from_components() {
-    size_t total_memory_reclaimed = 0;
+    size_t memory_reclaimed = 0;
```
Comment (denesb) [inline sstables/sstables.cc:1411]: Please add some motivation to the patches. It is ok for it to be short and just say that: in the next patch we want to use this to decide which component to reload, so keep the accounting in a member variable.

File: sstables/sstables.cc [hunk 2]
```diff
@@ -1429,6 +1429,19 @@ size_t sstable::total_memory_reclaimed() const {
     return _total_memory_reclaimed;
 }
 
+future<size_t> sstable::reload_reclaimed_components() {
+    if (_total_memory_reclaimed == 0) {
+        // nothing to reload
+        co_return 0;
+    }
+
+    _recognized_components.insert(component_type::Filter);
```
Comment (denesb) [inline sstables/sstables.cc:1438]: What will happen to reads attempting to probe the filter between this line and finishing the reload of the filter?

File: sstables/sstables_manager.cc [hunk 3]
```diff
@@ -169,6 +169,27 @@ void sstables_manager::increment_total_reclaimable_memory_and_maybe_reclaim(ssta
     smlogger.info("Reclaimed {} bytes of memory from SSTable components. Total memory reclaimed so far is {} bytes", memory_reclaimed, _total_memory_reclaimed);
 }
 
+future<> sstables_manager::maybe_reload_reclaimed_components() {
+    size_t memory_reclaim_threshold = _available_memory * _db_config.components_memory_reclaim_threshold();
+    size_t memory_available_for_reclaimable_components = memory_reclaim_threshold - _total_reclaimable_memory;
+
+    while (memory_available_for_reclaimable_components > 0) {
+        auto sstable_to_reload = std::lower_bound(_reclaimed.begin(), _reclaimed.end(), memory_available_for_reclaimable_components, [](const sstable& sst, size_t memory_available_for_reclaimable_components) {
```
Comment (denesb) [inline sstables/sstables_manager.cc:177]: I think this currently will try to reload the largest filters. I think we should reverse the sorting so we are loading the smallest filters first: this will maximize value, as we want to have the most filters in memory possible: so on evict we evict the largest (to minimize the amount of filters to evict) and on reload we reload the smallest (to maximize the amount of filters reloaded).

File: sstables/sstables_manager.cc [hunk 4]
```diff
@@ -194,6 +215,7 @@ void sstables_manager::remove(sstable* sst) {
     _undergoing_close.erase(_undergoing_close.iterator_to(*sst));
     delete sst;
     maybe_done();
+    maybe_reload_reclaimed_components().get();
```
Comment (denesb) [inline sstables/sstables_manager.cc:218]: What happens if there is a concurrent add() which wants to reclaim memory?

File: test/boost/sstable_datafile_test.cc [hunk 5]
```diff
@@ -3192,6 +3192,12 @@ SEASTAR_TEST_CASE(test_sstable_reclaim_memory_from_components) {
         BOOST_REQUIRE_EQUAL(sst_test.reclaim_memory_from_components(), total_reclaimable_memory);
         BOOST_REQUIRE_EQUAL(sst_test.total_reclaimable_memory_size(), 0);
         BOOST_REQUIRE_EQUAL(sst->filter_memory_size(), 0);
+
+        // Test sstable::reload_reclaimed_components() :
+        // Reloading should load the bloom filter back into memory
+        BOOST_REQUIRE_EQUAL(sst_test.reload_reclaimed_components(), total_reclaimable_memory);
+        BOOST_REQUIRE_EQUAL(sst_test.total_reclaimable_memory_size(), total_reclaimable_memory);
+        BOOST_REQUIRE_EQUAL(sst->filter_memory_size(), total_reclaimable_memory);
```
Comment (denesb) [inline test/boost/sstable_datafile_test.cc:3200]: This test is good to test the sstable API, but we need higher level tests, which add sstables to manager until we observe reclamation, then remove sstables, and observe previously evicted filters being reloaded.

**Review by denesb — CHANGES_REQUESTED** [review]
The approach looks good, but we need to consider concurrent operations and improve the tests. I also think we should reverse the reload order: load smallest filters first to maximize the amount we can reload.

File: test/boost/sstable_datafile_test.cc (hunk 5 above)
Comment (denesb) [reply test/boost/sstable_datafile_test.cc:3200]: We should also attempt to test concurrent evict/reload. I know this is hard to do reliably, maybe we can use failure injection API, or fuzzing.

**Comment by raphaelsc** [comment]
> The approach looks good, but we need to consider concurrent operations and improve the tests. I also think we should reverse the reload order: load smallest filters first to maximize the amount we can reload.

I haven't reviewed the patch, but one thing that I think we need to be careful is to make sure we're not loading and reclaiming back and forth, which can lead to waste of resources. This oscillation, which can cause filter mem usage to oscillate between the threshold could be caused by frequent bad estimations and compaction eventually correcting it. We need to apply some hysteresis.

File: sstables/sstables.cc (hunk 1 above)
Comment (lkshminarayanan) [reply sstables/sstables.cc]: Done.

File: test/boost/sstable_datafile_test.cc (hunk 5 above)
Comment (lkshminarayanan) [reply test/boost/sstable_datafile_test.cc:3206]: Done.

**Comment by lkshminarayanan** [comment]
> > The approach looks good, but we need to consider concurrent operations and improve the tests. I also think we should reverse the reload order: load smallest filters first to maximize the amount we can reload.
> 
> I haven't reviewed the patch, but one thing that I think we need to be careful is to make sure we're not loading and reclaiming back and forth, which can lead to waste of resources. This oscillation, which can cause filter mem usage to oscillate between the threshold could be caused by frequent bad estimations and compaction eventually correcting it. We may need to apply some hysteresis.

Yes this is a possibility. I have not addressed this yet as I first wanted to get the core functionality right. Maybe what we could do is if we observe reload is being called repeatedly on the same set of sstables, we could add an exponential backoff and delay the reload, giving sometime to the server to correct the situation. Or simply put a limit on how much times the same SSTable can be reloaded again.

File: sstables/sstables.hh [hunk 6]
```diff
@@ -593,6 +593,8 @@ private:
     // It is initialized to 0 to prevent the sstables manager from reclaiming memory
     // from the components before the SSTable has been fully loaded.
     mutable std::optional<size_t> _total_reclaimable_memory{0};
+    // Total memory reclaimed so far from this sstable
+    size_t _total_memory_reclaimed{0};
```
Comment (avikivity) [inline sstables/sstables.hh:602]: Don't know if we truly need to maintain this if we have a list of sstables. As long as we're under our limit, we can keep unreclaiming.

File: sstables/sstables.hh [hunk 7]
```diff
@@ -953,6 +957,13 @@ public:
     // Returns a read-only file for all existing components of the sstable
     future<std::unordered_map<component_type, file>> readable_file_for_all_components() const;
 
+    struct lesser_reclaimed_memory {
+        // comparator class to be used by the _reclaimed set in sstables manager
+        bool operator()(const sstable& sst1, const sstable& sst2) const {
+            return sst1.total_memory_reclaimed() < sst2.total_memory_reclaimed();
+        }
```
Comment (avikivity) [inline sstables/sstables.hh:966]: Is this sufficient? It's not a total order.

File: sstables/sstables_manager.cc [hunk 8]
```diff
@@ -169,6 +169,41 @@ void sstables_manager::increment_total_reclaimable_memory_and_maybe_reclaim(ssta
     smlogger.info("Reclaimed {} bytes of memory from SSTable components. Total memory reclaimed so far is {} bytes", memory_reclaimed, _total_memory_reclaimed);
 }
 
+future<> sstables_manager::maybe_reload_reclaimed_components() {
+    size_t memory_reclaim_threshold = _available_memory * _db_config.components_memory_reclaim_threshold();
+    size_t memory_available_for_reclaimable_components = memory_reclaim_threshold - _total_reclaimable_memory;
+
```
Comment (avikivity) [inline sstables/sstables_manager.cc:175]: This variable becomes out-of-date as soon as you co_await. Better to always compute from live data and not cache anything.

File: sstables/sstables_manager.cc [hunk 9]
```diff
@@ -169,6 +169,41 @@ void sstables_manager::increment_total_reclaimable_memory_and_maybe_reclaim(ssta
...
+        // Increment the total memory before reloading to prevent any parallel
+        // fibers from loading new bloom filters into memory.
+        _total_reclaimable_memory += reclaimed_memory;
+        _reclaimed.erase(sstable_to_reload);
+        // Use a lw_shared_ptr to prevent the sstable from getting deleted when
+        // the components are being reloaded.
+        auto sstable_ptr = sstable_to_reload->shared_from_this();
+        try {
+            memory_available_for_reclaimable_components -= co_await sstable_ptr->reload_reclaimed_components();
+        } catch (std::exception& e) {
```
Comment (avikivity) [inline sstables/sstables_manager.cc:195]: Better: catch (...) to not rely on inheritance from std::exception (and use std::current_exception())

File: sstables/sstables_manager.cc [hunk 10]
```diff
@@ -195,6 +230,9 @@ void sstables_manager::remove(sstable* sst) {
     _undergoing_close.erase(_undergoing_close.iterator_to(*sst));
     delete sst;
     maybe_done();
+    if (!_closing) {
+        (void)maybe_reload_reclaimed_components();
```
[the remark this answers is not in the document]
Comment (avikivity) [reply sstables/sstables_manager.cc:234]: A better (and standard) approach: have a fiber started by sstable_manager listen on a condition variable, and signal it during remove(). Wait for the fiber in sstable_manager::close(). This also prevents multiple maybe_reclaims running in parallel and thrashing shared state.

[the remark this answers is not in the document]
Comment (lkshminarayanan) [reply sstables/sstables_manager.cc]: Done. A new fiber now takes care of reloading the filter.

File: sstables/sstables_manager.cc (hunk 9 above)
Comment (lkshminarayanan) [reply sstables/sstables_manager.cc]: Done.

File: sstables/sstables.hh (hunk 7 above)
Comment (lkshminarayanan) [reply sstables/sstables.hh:966]: Yes, the boost [doc](https://www.boost.org/doc/libs/1_83_0/doc/html/intrusive/set_multiset.html) for intrusive set says that the `comparison functor must induce a strict weak ordering`. So, this should be sufficient.

File: sstables/sstables_manager.cc [hunk 11]
```diff
@@ -178,6 +183,39 @@ future<> sstables_manager::components_reloader_fiber() {
...
+            } catch (...) {
+                // reload failed due to some reason
+                sstlog.warn("Failed to reload reclaimed SSTable components : {}", std::current_exception());
+                // revert back changes made before the reload
+                _total_reclaimable_memory -= reclaimed_memory;
+                _reclaimed.insert(*sstable_to_reload);
+                break;
+            }
+
+            _total_memory_reclaimed -= reclaimed_memory;
```
Comment (denesb) [inline sstables/sstables_manager.cc:216]: Why this revert here? Is it to counter-balance the internal increment done by the sstable?

File: test/boost/sstable_datafile_test.cc [hunk 12]
```diff
@@ -3291,3 +3291,56 @@ SEASTAR_TEST_CASE(test_reclaimed_bloom_filter_deletion_from_disk) {
         BOOST_REQUIRE(!file_exists(filter_path).get());
     });
 }
+
+SEASTAR_TEST_CASE(test_bloom_filter_reclaim_during_reload) {
+    return test_env::do_with_async([] (test_env& env) {
+#ifndef SCYLLA_ENABLE_ERROR_INJECTION
+        fmt::print("Skipping test as it depends on error injection. Please run in mode where it's enabled (debug,dev).\n");
+#else
```
Comment (denesb) [inline test/boost/sstable_datafile_test.cc:3299]: Instead of `#else`, just add a `return` above to short-circuit the test when error injection is not available.

File: test/boost/sstable_directory_test.cc [hunk 13]
```diff
@@ -196,7 +196,7 @@ SEASTAR_TEST_CASE(sstable_directory_test_table_simple_empty_directory_scan) {
 // Test unrecoverable SSTable: missing a file that is expected in the TOC.
 SEASTAR_TEST_CASE(sstable_directory_test_table_scan_incomplete_sstables) {
     return sstables::test_env::do_with_async([] (test_env& env) {
-        auto sst = make_sstable_for_this_shard(std::bind(new_env_sstable, std::ref(env)));
+        auto sst = make_sstable_for_this_shard(std::bind(new_sstable, std::ref(env), env.tempdir().path().native(), generation_type(this_shard_id())));
```
Comment (denesb) [inline test/boost/sstable_directory_test.cc:199]: How did you discover this? Does this happen only with your changes in this PR? Seems unrelated.

**Review by denesb — COMMENTED** [review]
Approach looks good to me. Some minor comments.

File: test/boost/sstable_directory_test.cc (hunk 13 above)
Comment (lkshminarayanan) [reply test/boost/sstable_directory_test.cc:199]: yes, the failure occurs with this PR :
https://github.com/scylladb/scylladb/pull/18186#issuecomment-2099035844

File: sstables/sstables_manager.cc (hunk 11 above)
Comment (lkshminarayanan) [reply sstables/sstables_manager.cc:216]: yes, the current bloom filter is reloaded and not reclaimed anymore - so we decrement it here. If the same bloom filter is reclaimed twice - (i.e) reclaimed, then reloaded and then reclaimed sometime after that, the reclaimed memory would be counted twice without this decrement.

File: test/boost/sstable_datafile_test.cc (hunk 12 above)
Comment (lkshminarayanan) [reply test/boost/sstable_datafile_test.cc]: Done.

**Comment by denesb** [comment]
> ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1454/)

Seems to be related to the gossip fiasco from last week.
